hoodie-common/src/main/java/com/uber/hoodie/common/util/SpillableMapUtils.java

```diff
@@ -0,0 +1,155 @@
+import com.uber.hoodie.exception.HoodieCorruptedDataException;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.util.Optional;
+import java.util.zip.CRC32;
+
+public class SpillableMapUtils {
+
+  /**
+   * Using the schema and payload class, read value from disk and return a HoodieRecord
+   * @param file
+   * @param schema
+   * @param payloadClazz
+   * @param valuePosition
+   * @param valueLength
+   * @param <R>
+   * @return
+   * @throws IOException
+   */
+  public static <R> R read(RandomAccessFile file, Schema schema, String payloadClazz,
+                           long valuePosition, int valueLength) throws IOException {
+
+    DiskBasedMap.FileEntry fileEntry = readInternal(file, valuePosition, valueLength);
+    return (R) convertBytesToRecord(fileEntry.getValue(), schema, payloadClazz);
+  }
+
+  /**
+   * |crc|timestamp|sizeOfKey|SizeOfValue|key|value|
+   * @param file
+   * @param valuePosition
+   * @param valueLength
+   * @return
+   * @throws IOException
+   */
+  private static DiskBasedMap.FileEntry readInternal(RandomAccessFile file, long valuePosition, int valueLength) throws IOException {
+    file.seek(valuePosition);
+    long crc = file.readLong();
+    long timestamp = file.readLong();
+    int keySize = file.readInt();
+    int valueSize = file.readInt();
+    byte [] key = new byte[keySize];
+    file.read(key, 0, keySize);
+    byte [] value = new byte[valueSize];
+    if(!(valueSize == valueLength)) {
+      throw new HoodieCorruptedDataException("unequal size of payload written to external file");
+    }
+    file.read(value, 0, valueSize);
+    long crcOfReadValue = SpillableMapUtils.generateChecksum(value);
+    if(!(crc == crcOfReadValue)) {
+      throw new HoodieCorruptedDataException("checksum of payload written to external disk does not match");
+    }
+    return new DiskBasedMap.FileEntry(crc, keySize, valueSize, key, value, timestamp);
+  }
+
+  /**
+   * Write Value and other metadata necessary to disk. Each entry looks as follows:
+   *
+   * |crc|timestamp|sizeOfKey|SizeOfValue|key|value|
+   *
+   * @param outputStream
+   * @param fileEntry
+   * @return
+   * @throws IOException
+   */
+  public static long write(SizeAwareDataOutputStream outputStream, DiskBasedMap.FileEntry fileEntry) throws IOException {
+    return writeInternal(outputStream, fileEntry);
+  }
+
+
+  private static long writeInternal(SizeAwareDataOutputStream outputStream, DiskBasedMap.FileEntry fileEntry)
+      throws IOException {
+    outputStream.writeLong(fileEntry.getCrc());
+    outputStream.writeLong(fileEntry.getTimestamp());
+    outputStream.writeInt(fileEntry.getSizeOfKey());
+    outputStream.writeInt(fileEntry.getSizeOfValue());
+    outputStream.write(fileEntry.getKey());
+    outputStream.write(fileEntry.getValue());
+    outputStream.flush();
```

Comment:
do you always have to flush() ? it could be costly right?

Reply:
At the moment, SizeAwareDataOutputStream uses DataOutputStream (which uses FileOutputStream) which doesn't buffer anything, it flushes on every write(..) so flush() is a no-op. ( * The <code>flush</code> method of <code>OutputStream</code> does nothing). But we could consider using BufferedOutputStream and then flush only when a get(..) is called because get(..) uses a inputStream from the file so data needs to be written to disk.

Reply:
@vinothchandar here

Reply:
To clarify, I am talking about fsync/flush at Linux level (not particularly on java apis which could trigger them).. 


Given the entire task is going to be retried  if any failures happen, what I would suggest is to let Linux pdflush take care of flushing your appends to the log file completely.. Using a BufferedOutputStream may be tricky, if you need to do a get (like you mentioned) .. Most implementations keep a write buffer and decouple it from the flushing.. Flushing on every get will get expensive as well and may not actually help wuth anything, just incurring more system calls.. 

what do you think about the options?

Reply:
I agree with you. It seems like from this documentation : https://docs.oracle.com/javase/7/docs/api/java/io/OutputStream.html#flush() this will just incurr an additional system call. Although, my worry is that we perform get(..) after put(..) but I think the OS should manage that irrespective of this extra system call. Most implementations I've seen keep a BufferedOutputStream but in this case we can just rely on OS buffer and pdflush.

Reply:
does with or without flush() make a difference performance wise?  whats the AI here?